dipy/denoise/localpca.py

```diff
@@ -0,0 +1,114 @@
+            for j in range(patch_radius, arr.shape[1] - patch_radius, 1):
+                for i in range(patch_radius, arr.shape[0] - patch_radius, 1):
+
+                    X = np.zeros(
+                        (patch_size * patch_size * patch_size, arr.shape[3]))
+                    M = np.zeros(arr.shape[3])
+
+                    temp = arr[i - patch_radius: i + patch_radius + 1,
+                               j - patch_radius: j + patch_radius + 1,
+                               k - patch_radius: k + patch_radius + 1, :]
+                    X = temp.reshape(
+                        patch_size *
+                        patch_size *
+                        patch_size,
+                        arr.shape[3])
+                    # compute the mean and normalize
+                    M = np.mean(X, axis=0)
+                    X = X - np.array([M, ] * X.shape[0],
+                                     dtype=np.float64)
+
+                    # Compute the covariance matrix C = X_transpose X
+                    C = np.transpose(X).dot(X)
+                    C = C / X.shape[0]
+
+                    [d, W] = np.linalg.eigh(C)
+
+                    d[d < tou[i, j, k, :]] = 0
+
+                    W_hat = W * 0
+                    W_hat[:, d > 0] = W[:, d > 0]
+                    Y = X.dot(W_hat)
+                    X_est = Y.dot(np.transpose(W_hat))
+
+                    # theta value
+                    temp = X_est + \
+                        np.array([M, ] * X_est.shape[0], dtype=np.float64)
+                    temp = temp.reshape(
+                        patch_size, patch_size, patch_size, arr.shape[3])
+                    # Also update the estimate matrix which is X_est * theta
+
+                    theta[i - patch_radius: i + patch_radius + 1,
+                          j - patch_radius: j + patch_radius + 1,
+                          k - patch_radius: k + patch_radius + 1,
+                          :] = theta[i - patch_radius: i + patch_radius + 1,
+                                     j - patch_radius: j + patch_radius + 1,
+                                     k - patch_radius: k + patch_radius + 1,
+                                     :] + 1.0 / (1.0 + np.linalg.norm(d,
+                                                                      ord=0))
+
+                    thetax[i - patch_radius: i + patch_radius + 1,
+                           j - patch_radius: j + patch_radius + 1,
+                           k - patch_radius: k + patch_radius + 1,
+                           :] = thetax[i - patch_radius: i + patch_radius + 1,
+                                       j - patch_radius: j + patch_radius + 1,
+                                       k - patch_radius: k + patch_radius + 1,
+                                       :] + temp / (1 + np.linalg.norm(d,
+                                                                       ord=0))
+
+        denoised_arr = thetax / theta
```

Comment:
So, while my tracking computes, it seems they recompute the final noise variance on the residuals maps + correction factor, similar to the way it is computed on the smallest pca components at first. Since you now have a 4D local map of noise variance, they take the median at the end to get a robust 3D estimation.

For the debias step itself, it gets a tad more complicated as you need to compute a numerical inverse for the mapping function. So, eq. 5 works with the local SNR once again, this time using the computed denoised value and the new value from the residual noise map at each voxel.

This value of E(x)/sigma is then used in eq. 6 to build a table of values (or it can just be applied to each voxel, they do that to save some computation time here it seems).
------
And a new comment appeared between me starting and finishing this message, so here goes : 

It would be easier to get it included now as it is part of the algorithm. Since apparently when looking a difference of FA you see stuff along 1-2% difference, the value you can get out with and without bias correction is in the same order/higher if you move to high b-value/finer voxels (and thus less signal). 

Well anyway, let's see how it goes but all the parts seem to be in place already except for the one liner of eq 6. While it is true that you could use it for other functions, nlmeans work in the square domain and already include such a correction, so I don't know how applicable to other stuff it would be.